source/common/upstream/egds_cluster_mapper.cc

```diff
@@ -0,0 +1,177 @@
+  return active_monitors_.find(name) != active_monitors_.end();
+}
+
+void EgdsClusterMapper::addResource(absl::string_view name) {
+  if (active_monitors_.find(name) == active_monitors_.end()) {
+    active_monitors_.emplace(name, std::make_shared<ActiveEndpointGroupMonitor>(*this));
+    monitor_manager_.addMonitor(active_monitors_[name], name);
+  }
+}
+
+void EgdsClusterMapper::removeResource(absl::string_view name) {
+  if (active_monitors_.find(name) != active_monitors_.end()) {
+    monitor_manager_.removeMonitor(active_monitors_[name], name);
+    active_monitors_.erase(name);
+  }
+}
+
+void EgdsClusterMapper::ActiveEndpointGroupMonitor::update(
+    const envoy::config::endpoint::v3::EndpointGroup& group, absl::string_view version_info) {
+  if (group_.has_value() && Protobuf::util::MessageDifferencer::Equivalent(group_.value(), group)) {
+    ENVOY_LOG(info, "Resources {} are unchanged and do not need to be updated, version: {} ",
+              group.name(), version_info);
+    return;
+  }
+
+  group_ = group;
+  version_ = version_info.data();
+
+  // When the complete data of the cluster is initialized, subsequent updates to each EG are
+  // incrementally calculated.
+  if (parent_.cluster_initialized_) {
+    BatchUpdateHelper helper(*this);
+    parent_.delegate_.batchHostUpdateForEndpointGroup(helper);
+    return;
+  }
+
+  if (parent_.clusterDataIsReady()) {
+    // Initialize the cluster after all EG data is complete and the initialization is performed only
+    // once.
+    parent_.initializeCluster();
+  }
+
+  // Initialize the Endpoint Group, using the initial data as the baseline for subsequent
+  // calculations. Initialization does not trigger "updateHosts" notification.
+  initializeEndpointGroup();
+}
+
+void EgdsClusterMapper::ActiveEndpointGroupMonitor::initializeEndpointGroup() {
+  BatchUpdateHelper helper(*this, false);
+  EmptyBatchUpdateScope empty_update;
+  helper.batchUpdate(empty_update);
+}
+
+bool EgdsClusterMapper::ActiveEndpointGroupMonitor::BatchUpdateHelper::
+    calculateUpdatedHostsPerLocality(const uint32_t priority, const HostVector& new_hosts,
+                                     std::unordered_map<std::string, HostSharedPtr>& updated_hosts,
+                                     PriorityStateManager& priority_state_manager,
+                                     LocalityWeightsMap& new_locality_weights_map,
+                                     absl::optional<uint32_t> overprovisioning_factor) {
+  auto& host_set = parent_.priority_set_.getOrCreateMutableHostSet(priority);
+  ENVOY_LOG(debug,
+            "compute the updated host in the '{}' endpoint-group, added hosts count '{}', existed "
+            "hosts count '{}'",
+            parent_.group_.value().name(), new_hosts.size(), host_set.mutableHosts().size());
+  HostVector hosts_added;
+  HostVector hosts_removed;
+  const bool hosts_updated = parent_.parent_.cluster_.updateDynamicHostList(
```

Comment:
I believe this implies that each group update contains the full list of endpoints for this priority, which would limit the usefulness of groups to only define disjoint priorities (priorities cannot span groups). Is this correct/intentional? If not, how are changes to the same priority defined by multiple groups handled?

I made a comment on the design doc about this as well, can we make sure that we have clarity on the exact invariants imposed by the groups? I think it would be helpful to iron out these details before getting this PR reviewed

Reply:
Multiple groups that can define the same priority, can also be different, because activeEndpointGroupMonitor just calls the updateDynamicHostList function to compute the updated host (get added, deleted hosts), EdsCluster then uses this data to update its priority_set and call the original updateClusterPrioritySet function to complete the update.

Yes, it is necessary to discuss those details. I have replied in the design document.